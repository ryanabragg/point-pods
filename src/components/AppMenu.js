import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Drawer from '@material-ui/core/Drawer';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';

import MenuIcon from '@material-ui/icons/Menu';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import HomeIcon from '@material-ui/icons/Home';
import AddBoxIcon from '@material-ui/icons/AddBox';
import BallotIcon from '@material-ui/icons/Ballot';
import PeopleIcon from '@material-ui/icons/People';
import SettingsIcon from '@material-ui/icons/SettingsSharp';

const styles = theme => ({
  root: {
    width: '100%',
  },
  flex: {
    flexGrow: 1,
    flexShrink: 0,
    marginRight: theme.spacing.unit,
  },
  nowrap: {
    flexWrap: 'nowrap',
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20,
  },
  list: {
    width: 250,
  },
  spacer: {
    height: `${64 + 2 * theme.spacing.unit}px`,
    [theme.breakpoints.down('xs')]: {
      height: `${56 + 2 * theme.spacing.unit}px`,
    },
  },
  pad: {
    height: `${2 * theme.spacing.unit}px`,
  },
});

class AppMenu extends React.Component {
  state = {
    drawerOpen: false,
  };

  componentDidUpdate(prevProps, prevState) {
    if(this.props.usesBackIcon && this.props.usesBackIcon !== prevProps.usesBackIcon)
      this.setState({ drawerOpen: false });
  }

  toggleDrawer = (open) => () => {
    this.setState({
      drawerOpen: open,
    });
  };

  render() {
    const { classes, theme, position, color, title, toolbar, usesBackIcon } = this.props;

    return (
      <div className={classes.root}>
        <AppBar position={position} color={color}>
          <Toolbar className={classes.nowrap}>
            {usesBackIcon
              ? (
                <IconButton
                  aria-label='Back'
                  color='inherit'
                  className={classes.menuButton}
                  onClick={this.props.onBack}
                >
                  <ArrowBackIcon />
                </IconButton>
              ) : (
                <IconButton
                  aria-label='Menu'
                  color='inherit'
                  className={classes.menuButton}
                  onClick={this.toggleDrawer(true)}
                >
                  <MenuIcon />
                </IconButton>
              )
            }
            <Typography variant='title' color='inherit' className={classes.flex}>
              {title}
            </Typography>
            {toolbar}
          </Toolbar>
        </AppBar>
        {(position === 'fixed' || position === 'absolute')
          ? <div className={classes.spacer} />
          : <div className={classes.pad} />
        }
        <Drawer open={this.state.drawerOpen} onClose={this.toggleDrawer(false)}>
          <div
            tabIndex={0}
            role='button'
            onClick={this.toggleDrawer(false)}
            onKeyDown={this.toggleDrawer(false)}
          >
            <div className={classes.list}>
              <AppBar position='static'>
                <Toolbar>
                  <IconButton className={classes.menuButton} color='inherit' aria-label='Menu-Close'>
                    {theme.direction === 'rtl' ? <ChevronRightIcon /> : <ChevronLeftIcon />}
                  </IconButton>
                  <Typography variant='title' color='inherit'>
                    Menu
                  </Typography>
                </Toolbar>
              </AppBar>
              <List>
                <ListItem button component='a' href='/'>
                  <ListItemIcon>
                    <HomeIcon />
                  </ListItemIcon>
                  <ListItemText primary='Home' />
                </ListItem>
                <ListItem button component='a' href='/new'>
                  <ListItemIcon>
                    <AddBoxIcon />
                  </ListItemIcon>
                  <ListItemText primary='New Tournament' />
                </ListItem>
                <ListItem button component='a' href='/tournaments'>
                  <ListItemIcon>
                    <BallotIcon />
                  </ListItemIcon>
                  <ListItemText primary='Tournaments' />
                </ListItem>
                <ListItem button component='a' href='/players'>
                  <ListItemIcon>
                    <PeopleIcon />
                  </ListItemIcon>
                  <ListItemText primary='Players' />
                </ListItem>
              </List>
              <Divider />
              <List>
                <ListItem button component='a' href='/settings'>
                  <ListItemIcon>
                    <SettingsIcon />
                  </ListItemIcon>
                  <ListItemText primary='Settings' />
                </ListItem>
              </List>
            </div>
          </div>
        </Drawer>
      </div>
    );
  }
}

AppMenu.defaultProps = {
  position: 'fixed',
  color: 'primary',
  title: 'Point Pods',
  toolbar: null,
  usesBackIcon: false,
  onBack: () => null,
};

AppMenu.propTypes = {
  classes: PropTypes.object.isRequired, // added by withStyles
  position: PropTypes.oneOf(['static', 'fixed']),
  color: PropTypes.string,
  title: PropTypes.string,
  toolbar: PropTypes.element,
  usesBackIcon: PropTypes.bool,
  onBack: PropTypes.func,
};

export default withStyles(styles, { withTheme: true })(AppMenu);
