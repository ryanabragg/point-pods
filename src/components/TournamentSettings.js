import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import { withAPI } from '../api';

import Select from './Select';

import Divider from '@material-ui/core/Divider';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';

import AddIcon from '@material-ui/icons/Add';

const styles = theme => ({
  root: {
    margin: theme.spacing.unit,
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-evenly',
    alignItems: 'flex-start',
  },
  card: {
    overflow: 'visible',
    maxWidth: theme.breakpoints.values['sm'],
  },
  content: {
    padding: theme.spacing.unit,
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-evenly',
    alignItems: 'flex-start',
    alignContent: 'flex-start',
  },
  field: {
    minWidth: 250,
  },
  right: {
    marginLeft: 'auto',
  },
});

class TournamentSettings extends Component {
  state = {
    id: null,
    name: '',
    category: '',
    description: '',
    date: new Date().toJSON().slice(0,10),
    pairingMethod: '',
    pairingMethodInitial: '',
    podSizeMinimum: 0,
    podSizeMaximum: 0,
    pairingMethods: [],
    categories: [],
  };

  componentDidMount() {
    let loadState = {};
    this.props.api.Tournaments.all()
      .then(list => {
        Object.assign(loadState, {
          pairingMethods: Object.keys(this.props.api.pairingMethods)
            .map(key => ({
              label: key.slice(0,1).toUpperCase() + key.slice(1).toLowerCase(),
              value: this.props.api.pairingMethods[key],
            })),
          categories: list.map(tournament => tournament.category)
            .filter((cat, i, self) => cat && self.indexOf(cat) === i).sort(),
        });
        if(!this.props.id)
          return this.props.api.Settings.get()
            .then((settings) => Object.assign(loadState, settings));
        let found = list.filter(t => t.id == this.props.id);
        if(!found.length)
          throw new Error('ID not found');
        Object.assign(loadState, found[0]);
      })
      .then(() => this.setState(loadState))
      .catch(error => this.props.notification(error.message, 'error'));
  }

  handleInputChange = name => event => {
    let value = event.target.type === 'number'
      ? Number(event.target.value)
      : event.target.value;
    this.setState({
      [name]: value,
    });
  };

  handleSelectValue = name => (selected, action) => {
    if(selected === null)
      selected = {value: ''};
    this.setState((prevState, props) => {
      let val = typeof prevState[name] === 'number'
        ? Number(selected.value)
        : selected.value;
      return {
        [name]: val,
      };
    });
  };

  handleAddSelectValue = (name, list) => value => {
    this.setState((prevState, props) => {
      let val = typeof prevState[name] === 'number'
        ? Number(value)
        : value;
      return {
        [name]: val,
        [list]: [
          ...prevState[list],
          val,
        ],
      };
    });
  };

  handleSubmit = () => {
    if(!this.state.name)
      return;
    const values = Object.assign({}, this.state);
    if(values.id === null)
      delete values.id;
    this.props.api.Tournaments.set(values)
      .then(tournament => this.props.onSubmit(tournament))
      .catch(error => this.props.notification(error.message, 'error'));
  };

  render() {
    const { classes, title, autoFocus } = this.props;
    const {
      id,
      name,
      category,
      description,
      date,
      pairingMethod,
      pairingMethodInitial,
      podSizeMinimum,
      podSizeMaximum,
      pairingMethods,
      categories
    } = this.state;
    return (
      <div className={classes.root}>
        <Card className={classes.card}>
          {title && <CardHeader title={title} />}
          {title && <Divider />}
          <CardContent className={classes.content}>
            <TextField className={classes.field}
              autoFocus={autoFocus}
              margin='dense'
              id='tournament-name'
              label=' Name'
              type='text'
              value={name}
              onChange={this.handleInputChange('name')}
              fullWidth
            />
            <TextField className={classes.field}
              margin='dense'
              id='tournament-description'
              label='Description'
              type='text'
              value={description}
              onChange={this.handleInputChange('description')}
              fullWidth
            />
            <Select className={classes.field}
              id='tournament-category'
              label='Category'
              type='text'
              isClearable
              backspaceRemovesValue={false}
              isCreatable
              placeholder='Select or Type...'
              value={category && {label: category, value: category}}
              onChange={this.handleSelectValue('category')}
              options={categories.map(c => ({label: c, value: c}))}
              onCreateOption={this.handleAddSelectValue('category', 'categories')}
              getOptionLabel={option => option.label}
              getOptionValue={option => option.value}
            />
            <TextField className={classes.field}
              margin='dense'
              id='tournament-date'
              label='Date'
              type='date'
              value={date}
              onChange={this.handleInputChange('date')}
            />
            <Select className={classes.field}
              id='tournament-pairingMethod'
              label='Pairing Method'
              type='text'
              backspaceRemovesValue={false}
              value={pairingMethods.length && {label: pairingMethods.filter(m => m.value === pairingMethod)[0].label, value: pairingMethod}}
              onChange={this.handleSelectValue('pairingMethod')}
              options={pairingMethods}
              getOptionLabel={option => option.label}
              getOptionValue={option => option.value}
            />
            <Select className={classes.field}
              id='tournament-pairingMethodInitial'
              label='Initial Pairing Method'
              type='text'
              backspaceRemovesValue={false}
              value={pairingMethods.length && {label: pairingMethods.filter(m => m.value === pairingMethodInitial)[0].label, value: pairingMethodInitial}}
              onChange={this.handleSelectValue('pairingMethodInitial')}
              options={pairingMethods}
              getOptionLabel={option => option.label}
              getOptionValue={option => option.value}
            />
            <TextField className={classes.field}
              margin='dense'
              id='tournament-podSizeMinimum'
              label='Minimum Pod Size'
              type='number'
              value={podSizeMinimum}
              onChange={this.handleInputChange('podSizeMinimum')}
            />
            <TextField className={classes.field}
              margin='dense'
              id='tournament-podSizeMaximum'
              label='Maximum Pod Size'
              type='number'
              value={podSizeMaximum}
              onChange={this.handleInputChange('podSizeMaximum')}
            />
          </CardContent>
          <Divider />
          <CardActions>
            <Button className={classes.right}
              id='tournament-submit'
              aria-label={id ? 'Save' : 'Create'}
              onClick={this.handleSubmit}
            >
              {id ? 'Save' : 'Create'}
            </Button>
          </CardActions>
        </Card>
      </div>
    );
  }
}

TournamentSettings.defaultProps = {
  notification: (message, variant, duration, onClose) => null,
  id: null,
  onSubmit: (tournament) => null,
  title: '',
  autoFocus: false,
};

TournamentSettings.propTypes = {
  api: PropTypes.object.isRequired, // added by withAPI
  classes: PropTypes.object.isRequired, // added by withStyles
  notification: PropTypes.func,
  id: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  onSubmit: PropTypes.func,
  title: PropTypes.string,
  autoFocus: PropTypes.bool,
};

export default withStyles(styles, { withTheme: true })(withAPI(TournamentSettings));
