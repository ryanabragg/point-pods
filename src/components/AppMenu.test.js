import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Modal from '@material-ui/core/Modal';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';

import AppMenu from './AppMenu';
import TournamentSettings from './TournamentSettings';

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AppMenu />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches the prior snapshot', () => {
  const component = renderer.create(<AppMenu />);
  const toolbar = <div><button>test</button></div>;
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<AppMenu position='static' />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<AppMenu title='Test' toolbar={toolbar} usesBackIcon />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('content', () => {
  const toolbar = <div id='test' />;
  const component = shallow(<AppMenu />).dive().dive();
  expect(component.find(AppBar).at(0).prop('position')).toBe('fixed');
  expect(component.find(AppBar).at(1).prop('position')).toBe('static');
  expect(component.find(Toolbar).at(0).find(Typography).dive().dive().text()).toBe('Point Pods');
  expect(component.find(Toolbar).at(1).find(Typography).dive().dive().text()).toBe('Menu');
  expect(component.find('#test').length).toBe(0);
  component.setProps({
    position: 'static',
    color: 'secondary',
    title: 'Test',
    toolbar: toolbar,
  });
  expect(component.find(AppBar).at(0).prop('position')).toBe('static');
  expect(component.find(AppBar).at(0).prop('color')).toBe('secondary');
  expect(component.find(AppBar).at(1).prop('position')).toBe('static');
  expect(component.find(AppBar).at(1).prop('color')).toBe(undefined);
  expect(component.find(Toolbar).at(0).find(Typography).dive().dive().text()).toBe('Test');
  expect(component.find(Toolbar).at(1).find(Typography).dive().dive().text()).toBe('Menu');
  expect(component.find('#test').length).toBe(1);
});

test('back button', () => {
  const spy = jest.fn(v => v);
  const component = shallow(<AppMenu />).dive().dive();
  expect(component.find(IconButton).at(0).prop('aria-label')).toBe('Menu');
  expect(component.find(IconButton).at(1).prop('aria-label')).toBe('Menu-Close');
  component.setProps({
    usesBackIcon: true,
    onBack: spy,
  });
  expect(component.find(IconButton).at(0).prop('aria-label')).toBe('Back');
  expect(component.find(IconButton).at(1).prop('aria-label')).toBe('Menu-Close');
  component.find(IconButton).at(0).simulate('click');
  expect(spy.mock.calls.length).toBe(1);
});

test('menu drawer', () => {
  const spy = jest.fn(v => v);
  const component = shallow(<AppMenu />).dive().dive();
  expect(component.find(Modal).length).toBe(1);
  expect(component.state('drawerOpen')).toBe(false);
  component.find(IconButton).at(0).simulate('click');
  expect(component.state('drawerOpen')).toBe(true);
  expect(component.find(ListItemText).length).toBe(5);
  expect(component.find(ListItemText).at(0).prop('primary')).toBe('Home');
  expect(component.find(ListItemText).at(1).prop('primary')).toBe('New Tournament');
  expect(component.find(ListItemText).at(2).prop('primary')).toBe('Tournaments');
  expect(component.find(ListItemText).at(3).prop('primary')).toBe('Players');
  expect(component.find(ListItemText).at(4).prop('primary')).toBe('Settings');
  expect(component.state('modalOpen')).toBe(false);
  expect(component.find(Modal).prop('open')).toBe(false);
  component.find(ListItem).at(1).simulate('click');
  expect(component.state('modalOpen')).toBe(true);
  expect(component.find(Modal).prop('open')).toBe(true);
});

test('new tournament modal', () => {
  const spy = jest.fn(v => v);
  const component = shallow(<AppMenu />).dive().dive();
  expect(component.find(Modal).find(TournamentSettings).length).toBe(1);
  expect(component.state('modalOpen')).toBe(false);
  expect(component.find(Modal).prop('open')).toBe(false);
  component.find(ListItem).at(1).simulate('click');
  expect(component.state('modalOpen')).toBe(true);
  expect(component.find(Modal).prop('open')).toBe(true);
  component.setProps({ onCancelNew: spy });
  component.find(TournamentSettings).find(Button).at(0).simulate('click');
  expect(component.state('modalOpen')).toBe(false);
  expect(component.find(Modal).prop('open')).toBe(false);
  expect(spy.mock.calls.length).toBe(1);
  component.setProps({ showNew: true });
  expect(component.state('modalOpen')).toBe(false);
  expect(component.find(Modal).prop('open')).toBe(true);
});
