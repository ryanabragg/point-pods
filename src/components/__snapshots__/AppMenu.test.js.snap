// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the prior snapshot 1`] = `
<div
  className=""
>
  <header
    className="MuiPaper-root-134 MuiPaper-elevation4-140 MuiAppBar-root-126 MuiAppBar-positionFixed-127 MuiAppBar-colorPrimary-132 mui-fixed"
  >
    <div
      className="MuiToolbar-root-161 MuiToolbar-regular-163 MuiToolbar-gutters-162 APIComponent-nowrap-120"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-171 MuiIconButton-root-165 MuiIconButton-colorInherit-166 APIComponent-menuButton-121"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-170"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-174"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-181"
        />
      </button>
      <h2
        className="MuiTypography-root-188 MuiTypography-title-194 MuiTypography-colorInherit-207 APIComponent-flex-119"
      >
        Point Pods
      </h2>
    </div>
  </header>
  <div
    className="APIComponent-spacer-123"
  />
</div>
`;

exports[`matches the prior snapshot 2`] = `
<div
  className=""
>
  <header
    className="MuiPaper-root-134 MuiPaper-elevation4-140 MuiAppBar-root-126 MuiAppBar-positionStatic-130 MuiAppBar-colorPrimary-132"
  >
    <div
      className="MuiToolbar-root-161 MuiToolbar-regular-163 MuiToolbar-gutters-162 APIComponent-nowrap-120"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-171 MuiIconButton-root-165 MuiIconButton-colorInherit-166 APIComponent-menuButton-121"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-170"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-174"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-181"
        />
      </button>
      <h2
        className="MuiTypography-root-188 MuiTypography-title-194 MuiTypography-colorInherit-207 APIComponent-flex-119"
      >
        Point Pods
      </h2>
    </div>
  </header>
  <div
    className="APIComponent-pad-124"
  />
</div>
`;

exports[`matches the prior snapshot 3`] = `
<div
  className=""
>
  <header
    className="MuiPaper-root-134 MuiPaper-elevation4-140 MuiAppBar-root-126 MuiAppBar-positionFixed-127 MuiAppBar-colorPrimary-132 mui-fixed"
  >
    <div
      className="MuiToolbar-root-161 MuiToolbar-regular-163 MuiToolbar-gutters-162 APIComponent-nowrap-120"
    >
      <button
        aria-label="Back"
        className="MuiButtonBase-root-171 MuiIconButton-root-165 MuiIconButton-colorInherit-166 APIComponent-menuButton-121"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-170"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-174"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-181"
        />
      </button>
      <h2
        className="MuiTypography-root-188 MuiTypography-title-194 MuiTypography-colorInherit-207 APIComponent-flex-119"
      >
        Test
      </h2>
      <div>
        <button>
          test
        </button>
      </div>
    </div>
  </header>
  <div
    className="APIComponent-spacer-123"
  />
</div>
`;
